File: cqp.R
# function to pass command as string to a headless cqp process and skip first
# line if it contains version number
query_cqp <- function(query) {
  out <- system2("echo",
    paste("'", query, "'", "| cqp -c"),
    stdout = TRUE
  )
  if (grepl("CQP version", out[1]))
    return(out[-1]) else return(out)
}

# save CQP command(s) as string, cqp -e convenience features are off, e.g.,
# every statement has to be delimited with `;`
# `cat` has to be called explicitely
query <- r"(
  BROWN;
  query = "test";
  tabulate Last match[-5]..match[5] word;
)"

# can be saved directly as variable or array
concordance <- query_cqp(query)
concordance

# save as data.frame
read.table(text = query_cqp(query), sep = " ", quote = "")
# TODO: cwbwrapr

# automation with string interpolation by `glue` library
library(glue)
library(cwbwrapr)

corpora <- c("BNC-BABY", "BROWN")

concordances <- lapply(corpora, \(corpus) {
    query <- glue(r"(
      {corpus};
      query = "test";
      tabulate Last match[-5]..match[5] word;
    )")
    query_cqp(query)
})

# get counts of `hw` for BNC, and `lemma` for BROWN
corpora <- c("BNC-BABY", "BROWN")
p_attrs <- c("hw", "lemma")

frequency_lists <- mapply(\(corpus, p_attr) {
    query <- glue(r"(
      set PrettyPrint no;
      {corpus};
      query=[class = "ADV"];
      count query by {p_attr};
    )")
    out <- query_cqp(query)
    cwbwrapr::read_freqs(text = out) # FIXME: path is missing
}, corpora, p_attrs)

File: cqp.py
#!/usr/bin/env python

# Possibly an alternative:
# pip install cwb-ccc
#
# from ccc import Corpora ...
#
# https://github.com/ausgerechnet/cwb-ccc#installation
# didn't manage to install on my system

import subprocess
import pandas as pd

# function to pass command as string to a headless cqp process
# and read into a pandas data.frame
# set `skiprows` to 0 if your version of CQP doesn't print the version nr.
def query_cqp(
    query,
    header=None,
    sep="\s+",     # or \t+ depending on output
    quoting=3,   # disable quoting
    skiprows=1,
    **kwargs
):
    cmd = f"echo '{query}' | cqp -c"
    with subprocess.Popen(cmd, shell=True, stdout = subprocess.PIPE) as proc:
        return pd.read_table(
            proc.stdout,
            header=header,
            sep=sep,
            quoting=quoting,
            skiprows=skiprows,
            **kwargs
        )

  # if "CQP version" in out[0]:
  #   return(out[-1]) else return(out)


# save CQP command(s) as string, cqp -e convenience features are off, e.g.,
# every statement has to be delimited with `;`
# `cat` is not called implicitely (in non-interactive mode `tabulate` is
# preferable to cat)
query = """
  BROWN;
  query = "test";
  tabulate Last match[-5]..match[5] word;
"""

concordance = query_cqp(query)
print(concordance)

each_corpus = lambda corpus : f"""
    {corpus};
    query = "test";
    tabulate Last match[-5]..match[5] word;
"""

corpora = ("BNC-BABY", "BROWN")
queries = map(each_corpus, corpora)
concordances = map(query_cqp, queries)

for conc in concordances:
    print(conc[:15])

# get counts of `hw` for BNC, and `lemma` for BROWN
corpora = ("BNC-BABY", "BROWN")
p_attrs = ("hw", "lemma")

each_corpus_lemma = lambda corpus, p_attr : f"""
    set PrettyPrint no;
    {corpus};
    query=[class = "ADV"];
    count query by {p_attr};
"""

queries = map(each_corpus_lemma, corpora, p_attrs)

# `usecols` to skip offset (2nd column in count output)
frequency_lists = (query_cqp(
    query,
    usecols=[0, 2],
    names = ["f", "type"]
) for query in queries)

for corpus, freq in zip(corpora, frequency_lists):
    print(f"corpus: {corpus}\n", freq[:15], "\n")



# In older versions without f-strings, the lambdas look like this:
# each_corpus = lambda corpus : """
#     %s;
#     query = "test";
#     tabulate Last match[-5]..match[5] word;
# """ % corpus

# each_corpus_lemma = lambda corpus, p_attr : """
#     set PrettyPrint no;
#     %s;
#     query=[class = "ADV"];
#     count query by %s;
# """ % (corpus, p_attr)

